File: 1723.find-minimum-time-to-finish-all-jobs.py
#
# @lc app=leetcode id=1723 lang=python3
#
# [1723] Find Minimum Time to Finish All Jobs
#

# @lc code=start
from functools import lru_cache
from typing import List


class Solution:
    '''
    Very similar problem to Leetcode 698 Partition to K Equal Sum Subsets
    Again we sort the jobs from large to small and states from small to large

    Key difference: we do not have target here. Instead the DFS function should return the minimum time to finish instad of whether can form target.
    '''
    @lru_cache(None)
    def dfs(self, i, state):
        stateList = [int(c) for c in state.split("_")]
        if i == len(self.Nums):
            return max(stateList)
        else:
            result = float('Inf')
            for l in range(self.k):
                stateList[l] += self.Nums[i]
                if max(stateList) < result:
                    result = min(result, self.dfs(i + 1, "_".join([str(j) for j in sorted(stateList)])))
                stateList[l] -= self.Nums[i]
            return result

    def minimumTimeRequired(self, jobs: List[int], k: int) -> int:
        self.Nums = sorted(jobs, reverse=True)
        self.k = k
        return self.dfs(0, "_".join(['0']*k))
        
# @lc code=end

File: 1723.find-minimum-time-to-finish-all-jobs.cs
/*
 * @lc app=leetcode id=1723 lang=csharp
 *
 * [1723] Find Minimum Time to Finish All Jobs
 */

// @lc code=start
using System.Collections.Generic;
using System;
using System.Linq;
public class Solution 
{
    Dictionary<Tuple<int, string>, int> cache = new Dictionary<Tuple<int, string>, int>();
    int[] Nums;
    int K;
    public int DFS(int i, string state)
    {
        Tuple<int, string> t = new Tuple<int, string>(i, state);
        int result;
        if (cache.ContainsKey(t))
        {
            return cache[t];
        }
        else if (i == Nums.Length)
        {
            result = state.Split('_').Select(s => Int32.Parse(s)).Max();                
        }
        else
        {
            int[] stateList = state.Split('_').Select(s => Int32.Parse(s)).ToArray();
            result = Int32.MaxValue;
            for (int l = 0; l < K; l++)
            {
                stateList[l] += Nums[i];
                if (stateList.Max() < result)
                {
                    result = Math.Min(result, DFS(i + 1, string.Join("_", stateList.OrderBy(j => j).Select(j => j.ToString()))));
                }
                stateList[l] -= Nums[i];
            }
        }
        cache.Add(t, result);
        return result;
    }

    public int MinimumTimeRequired(int[] jobs, int k) 
    {
        Nums = jobs.OrderByDescending(n => n).ToArray();
        K = k;
        return DFS(0, string.Join("_", Enumerable.Repeat("0", k)));            
    }
}
// @lc code=end

